File: src/curves/secp256k1.zig
// src/curves/secp256k1.zig — secp256k1 curve parameters and thin wrappers.
// y² = x³ + 7 (mod P)  — A=0, B=7

const utils = @import("utils.zig");

pub const P = @as(u256, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F);
pub const N = @as(u256, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141);
pub const A: u256 = 0;
pub const B: u256 = 7;
pub const Gx = @as(u256, 0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798);
pub const Gy = @as(u256, 0x483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8);
pub const G = utils.Point{ .x = Gx, .y = Gy };
pub const HALF_N = N >> 1;

pub fn pointAdd(a: ?utils.Point, b: ?utils.Point) ?utils.Point {
    return utils.pointAdd(a, b, P, A);
}

pub fn pointMultiply(scalar: u256, point: utils.Point) ?utils.Point {
    return utils.pointMultiply(scalar, point, P, A, N);
}

pub fn pointMultiplyG(scalar: u256) ?utils.Point {
    return utils.pointMultiply(scalar, G, P, A, N);
}

pub fn isOnCurve(pt: utils.Point) bool {
    return utils.isOnCurve(pt, P, A, B);
}

pub fn deterministicK(priv_bytes: *const [32]u8, hash_bytes: *const [32]u8, attempt: u8, out: *[32]u8) void {
    utils.deterministicK(priv_bytes, hash_bytes, attempt, N, out);
}

pub fn hashToScalar(seed: []const u8, label: []const u8, out: *[32]u8) void {
    utils.hashToScalar(seed, label, N, out);
}

File: src/curves/secp256k1.js
import shim from "../shim.js";
import base62 from "../base62.js";
import settings from "../settings.js";
import aeskey from "../aeskey.js";
import sha256 from "../sha256.js";
import hash from "../hash.js";
import createCurveCore from "./utils.js";

const P = BigInt(
  "0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F",
);
const N = BigInt(
  "0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141",
);
const A = 0n;
const B = 7n;
const G = {
  x: BigInt(
    "0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798",
  ),
  y: BigInt(
    "0x483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8",
  ),
};
const core = createCurveCore({
  curve: "secp256k1",
  P,
  N,
  A,
  B,
  G,
  shim,
  base62,
  settings,
  sha256,
  extras: { aeskey, hash },
});

export default core;
